Support for custom FHIR profile

chandrashekar-s opened the pull request:
## Description of what I changed

- The user can load the definitions of custom profiles in a directory and the application will load these structure definitions during launch.
- The application overrides the default profiles for the resources with the ones loaded by the user. Currently, only one custom profile is supported for a resource type. In case if a custom profile exists for a resource type, then it will take precedence over the default profile. In case of multiple custom profiles for the same resource type, the custom profile url that is lexicographically smaller will be chosen.
- While converting the FHIR resources to Avro records, the overloaded profiles will be applied on the FHIR resources irrespective of which FHIR profile they belong to.

Issues Fixed:
----------
#560
#400 
#753 
#558 

Pending Changes:
------------------
In the long term, we need to provide support for multiple custom profiles. This is being tracked in #980 
The issues mentioned in #960 and #961 are partially fixed in the current PR.

## E2E test

The application was loaded with custom `us-core-profile` definitions and the  pipeline was run. The tables that were created in the spark thrift server contained the extended definitions defined in the `us-core-profile`.

![image](https://github.com/google/fhir-data-pipes/assets/29541718/fc90a552-21b1-4fd7-bab4-2aa276388337)

TESTED:

Please replace this with a description of how you tested your PR beyond the
automated e2e/unit tests.

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

chandrashekar-s commented:
@bashir2 The changes have been tested for the `Patient` resource with `us-core-profile`. For other resources it was failing because of existing issues in bunsen, will file a bug for the same. Also, for supporting multiple profiles will file another bug.

bashir2 commented on pipelines/common/src/main/java/com/google/fhir/analytics/AvroUtil.java:
```diff
@@ -0,0 +1,68 @@
+    if (!map.containsKey(resourceType)) {
+      Map<String, String> mapping =
+          FhirContexts.resourceProfileMappingsFor(fhirContext.getVersion().getVersion());
+      String profile = mapping.get(resourceType);
```
Can `profile` be ever `null` here?

bashir2 reviewed (commented):
Thanks @chandrashekar-s for this change; this significantly simplifies using extensions/profiles. Many of my comments are minor/questions. But there is a fairly large suggestion to simplify `ProfileProvider` API and related comments re. what happens if there are conflicting profiles. BTW, sorry that it took too long for the review; I meant to do this last week but did not manage to finish it until now. I also need to have a deeper look at the tests which I will do after addressing the major comments.

chandrashekar-s reviewed (commented):
Thanks @bashir2 for the changes. I have simplified the APIs and created 2 new classes namely `ProfileMapperFhirContexts` and `ProfileMapperProvider` which contains most of the logic to load and create mappings for profiles.

Also, I have made changes to fix this [issue](https://github.com/google/fhir-data-pipes/issues/400), to make fhirVersion configurable.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/FhirContexts.java:
```diff
@@ -72,4 +154,28 @@ public static FhirContext forR4() {
+   *
+   * @param fhirVersionEnum the fhir version for which the details are to be removed
+   */
+  public static void deRegisterFhirContext(FhirVersionEnum fhirVersionEnum) {
```
Yeah, I think it makes sense. These were needed for testing, otherwise during `mvn clean install` the test cases failing because the profiles loaded as part of the previous test classes was being reused in the current test class and was throwing conflict. I have annotated the method with `@VisibleForTesting` and left it for now. Do you have any other suggestions?

chandrashekar-s commented on bunsen/bunsen-uscore-resources/src/main/java/com/cerner/bunsen/UsCoreProfileProvider.java:
```diff
@@ -2,186 +2,534 @@
       PrePopulatedValidationSupport support, FhirContext context) {
+    Map<String, String> mapping = new HashMap<>();
     IParser parser = context.newJsonParser();
     // TODO remove non-StructureDefinition files in definitions-stu3/ if they are not needed.
```
Undid the changes in this class. Created `ProfieMapperProviderImplTest` as an example on how to load custom profiles. Need to add more test cases though.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/profiles/BaseProfileProvider.java:
```diff
@@ -0,0 +1,118 @@
+
+      if (fhirContext.getResourceTypes().contains(type)
+          && baseDefinition.startsWith("http://hl7.org/fhir/StructureDefinition/")) {
+        resourceProfileMap.put(type, url);
```
Done. Doing it in `ProfileMapperProviderImpl` now.

bashir2 commented on pipelines/controller/src/main/java/com/google/fhir/analytics/PipelineManager.java:
```diff
@@ -166,10 +171,17 @@ private void setLastRunStatus(LastRunStatus status) {
     initialiseFlinkConfiguration();
+    fhirContext =
+        ProfileMapperFhirContexts.getInstance()
+            .contextFor(dataProperties.getFhirVersion(), dataProperties.getProfileDefinitionsDir());
```
Inside the next call for initializing ` avroConversionUtil`, we have another call of `ProfileMapperFhirContexts.contextFor` and they should be consistent; this is not great. I think under the `pipelines` module, there should be only one [singleton] class that is aware of ProfileMapperFhirContexts existence; everyone else should use that class for getting `FhirContext` (which is initialized with the right version and right profile-def-dir). That class can be `AvroConversionUtil` but regardless let's avoid this implicit consistency requirements (e.g., between `fhirContext` and `avroConversionUtil`) and enforce them in code.

chandrashekar-s replied:
Yes this makes sense. I have made changes so that only `AvroConversionUtil` is aware of the `ProfileMapperFhirContexts` and everyone should use `AvroConversionUtil` instance to fetch `FhirContext`.

bashir2 reviewed (commented):
Thanks @chandrashekar-s for the changes; I think the APIs are becoming more consistent but we are not there yet. I have highlighted some of the implicit consistency requirements, e.g., between `FhirContext`, `ProfileMapperFhirContexts` and `AvroConversionUtil` and suggested ways to enforce these by more encapsulation.

There is also another high-level issue that I am not sure where to comment about: I was expecting that we will remove some of the old profile handling code in this PR, e.g., `UsCoreProfileProvider`; why aren't those included in this PR?

chandrashekar-s reviewed (commented):
Thanks @bashir2 for providing the detailed feedback.

chandrashekar-s commented on pipelines/controller/config/application.yaml:
```diff
@@ -134,6 +134,13 @@ fhirdata:
+  # supported. Currently only one profile is supported for any given resource type.
+  profileDefinitionsDir: "config/profile-definitions"
+
+  # The fhir version to be used for the FHIR Context APIs
```
Done

chandrashekar-s commented on pipelines/controller/config/profile-definitions/StructureDefinition-us-core-birthsex.json:
```diff
@@ -0,0 +1,327 @@
+{
+  "resourceType": "StructureDefinition",
```
If we package all these definitions in one module and create a jar out of it, then the structure definition files have to be read as `classpath` files, whereas currently the `ProfileMapperFhirContexts.contextFor(FhirVersionEnum fhirVersion, @Nullable String structureDefinitionsPath)` function expects files from a directory instead of classpath. This is done this way so that the user can configure definitions in a directory during deployment.

chandrashekar-s replied:
provided classpath support and have removed the definitions from the `pipelines`.

chandrashekar-s commented on pipelines/common/src/main/java/com/google/fhir/analytics/AvroConversionUtil.java:
```diff
@@ -49,9 +52,9 @@ public class AvroConversionUtil {
 
-  // TODO make this map FHIR version dependent: https://github.com/google/fhir-data-pipes/issues/400
-  private final Map<String, AvroConverter> converterMap;
+  private final Map<FhirVersionEnum, Map<String, AvroConverter>> converterMap;
```
Currently the application can support only one `fhirVersion` at a time. Yes, your point is valid and I have made changes according to the inputs you have given.

chandrashekar-s commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/R4AvroConverterUsCoreTest.java:
```diff
@@ -559,239 +569,6 @@ public void testCompile() throws IOException {
     }
   }
 
-  // TODO add test profile for R4: https://github.com/google/fhir-data-pipes/issues/558
```
These tests have been uncommented and moved to [R4AvroConverterCustomProfileTest.java](https://github.com/google/fhir-data-pipes/pull/924/commits/edbba19c45f88f5d273720432c3a0b0a92e4e36f#diff-b988098f40473a9c057d544accfe28686ae807b135d8167dea47c2d96851af14)

chandrashekar-s replied:
Yes, this should fix the issue #558. However, the test cases for both of them are identical now. We should also write some unit tests which are specific to each version, example testing the `R4StructureDefinitions.getReferenceTargetProfiles()` method

chandrashekar-s commented:
@bashir2 In the latest commit, I have added support for configuring fhir profiles via classpath names. By default, the classpath name containing the `us-core` fhir profiles have been configured, but user can override them via a different directory as well.

Also, all the `us-core` fhir profiles definitions have been moved to a common package and is being referenced in all the test cases now.

There are few issues which are open and are being tracked via #960 and #961.

bashir2 commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMapperFhirContexts.java:
```diff
@@ -0,0 +1,179 @@
+      String errorMsg =
+          String.format(
+              "The fhirContext has not been initialised yet for version=%s, Please initialise the"
+                  + " fhirContext using the method contextFor(FhirVersionEnum fhirVersion,"
```
nit:
```suggestion
                  + " fhirContext using one of the flavors of contextFor() methods"
```

chandrashekar-s replied:
Done

bashir2 commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMapperFhirContexts.java:
```diff
@@ -0,0 +1,179 @@
+    if (fhirContextData != null) {
+      isContextLoadedWithDifferentConfig(fhirContextData, structureDefinitionsPath, isClasspath);
+    } else {
+      FhirContext context = new FhirContext(fhirVersion);
```
I understand we have guards against recreating `FhirContext` objects, but still should we use `FhirContext.forCached()` method (in HAPI) instead? I mean are there any drawbacks?

chandrashekar-s replied:
I don't see a drawback, we can use this function to create a single FhirContext for the given version. However, I am not changing the current code, because I need to load different StructureDefinitions for different Test and by static context I am not able to load the second ones.

bashir2 replied:
Okay then if this is required, please feel free to resolve this comment (maybe also add a single line comment before this why we are not using the cached version).

bashir2 commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMappingProvider.java:
```diff
@@ -0,0 +1,250 @@
+        context.getValidationSupport().fetchAllStructureDefinitions();
+    for (IBaseResource definition : defaultDefinitions) {
+      support.addStructureDefinition(definition);
+      // Links the profile only if the definition belongs to a base resource
```
nit: Consider documenting why we need the following check (the same explanation you provide to my question before).

chandrashekar-s replied:
Done

bashir2 reviewed (commented):
Thanks @chandrashekar-s for the changes; I think the APIs are much cleaner. I am not done with my next review round yet but I wanted to share the comments I have so far (not to delay this too much). In particular, I still need to think more about the one extension per resource type and especially the changes you needed to do in us-core.

Also as I mentioned in one of the comments, I feel that this PR is bringing in too much StructureDefinition JSON files (e.g., SDC ones) some of which are only for testing purposes (IIUC), so it would be nice if we can reduce that.

chandrashekar-s commented on bunsen/bunsen-core-r4/src/test/java/com/cerner/bunsen/r4/TestData.java:
```diff
@@ -404,116 +445,121 @@ public static Provenance newProvenance() {
-
-  //   return patient;
-  // }
+  public static Patient newBunsenTestProfilePatient() {
```
Done

chandrashekar-s commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/R4AvroConverterCustomProfileTest.java:
```diff
@@ -0,0 +1,373 @@
+
+public class R4AvroConverterCustomProfileTest {
+
+  // TODO add test profile for R4: https://github.com/google/fhir-data-pipes/issues/558
```
Yes, this has fixed the R4 extensions test issues. So removing the TODO.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/BasePipelineOptions.java:
```diff
@@ -45,6 +45,14 @@ public interface BasePipelineOptions extends PipelineOptions {
 
+  @Description(
+      "Classpath containing the structure definition files for any custom profiles that needs to be"
+          + " supported")
```
Done

chandrashekar-s commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/Stu3AvroConverterCustomProfileTest.java:
```diff
@@ -0,0 +1,369 @@
+  private static Patient testBunsenTestProfilePatientDecoded;
+
+  @BeforeClass
+  public static void setUp() throws URISyntaxException, ProfileMapperException {
```
Yes, this has to be done and only include specific test cases in the relavant files. Example test cases for the function `R4StructureDefinitions.getReferenceTargetProfiles()`.

bashir2 commented:
Thanks @chandrashekar-s for the changes; is this ready for another round of review? I mean, are the changes we discussed last week applied?

chandrashekar-s commented:
> Thanks @chandrashekar-s for the changes; is this ready for another round of review? I mean, are the changes we discussed last week applied?

There are some minor documentation changes pending, will update once done (should be done today).

chandrashekar-s commented:
@bashir2 In the latest commit, I have partially addressed the issues listed in #960 and #961. I have also created the issue #980 to create a generic solution for multiple extensions.

chandrashekar-s commented:
> @bashir2 In the latest commit, I have partially addressed the issues listed in #960 and #961. I have also created the issue #980 to create a generic solution for multiple extensions. Updated the description of the PR to reflect the current implementation status.
> 
> You can review the last 2 commits now.

The description also contains a Pending Items section which will be addressed later on, not in this PR.

bashir2 commented on bunsen/bunsen-core/pom.xml:
```diff
@@ -38,7 +38,20 @@
-
+    <dependency>
+      <groupId>ca.uhn.hapi.fhir</groupId>
+      <artifactId>hapi-fhir-validation-resources-r4</artifactId>
```
Don't we require this and the next dep in the production code too? I mean why just `test` scope? If these are only needed for `test` scope, then for example, should we also update `hapi-fhir-structures-dstu3` above to `test` scope as well?

bashir2 commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/R4AvroConverterCustomProfileTest.java:
```diff
@@ -0,0 +1,363 @@
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class R4AvroConverterCustomProfileTest {
```
nit: Consider adding a short comment to describe the general plan in these tests (e.g., which profiles are involved and how is it different from R4 and STU3 versions.

chandrashekar-s replied:
Added comments, to clarify these are just some additional test cases for user defined profiles.

bashir2 commented on bunsen/extension-structure-definitions/src/main/resources/stu3-us-core-definitions/StructureDefinition-us-core-ethnicity.json:
```diff
@@ -0,0 +1 @@
+{"resourceType":"StructureDefinition","id":"us-core-ethnicity","text":{"status":"generated","div":"<div xmlns=\"http://www.w3.org/1999/xhtml\"><table border=\"0\" cellpadding=\"0\" cellspacing=\"0\" style=\"border: 0px #F0F0F0 solid; font-size: 11px; font-family: verdana; vertical-align: top;\"><tr style=\"border: 1px #F0F0F0 solid; font-size: 11px; font-family: verdana; vertical-align: top;\"><t ...
```
It seems that we have two versions of this file, one here and one under `other-profile-definitions`, is that right/needed? In general what is the purpose of the `other-profile-definitions`? As we discussed, I think we should try not to duplicate SD files from US-Core and keep only one version in our repo if possible.

chandrashekar-s replied:
Agreed. Removed this SD definition from the `other-profile-definitions` directory. `other-profile-definitions` contains additional profile defined by the user manually. It is just used for additional testing alone. I have removed this directory from the module `extension-structure-definitions` and moved to the `test/resources` directory of `bunsen-avro`.

bashir2 commented on bunsen/extension-structure-definitions/src/main/resources/stu3-us-core-definitions/StructureDefinition-us-core-race.json:
```diff
@@ -0,0 +1 @@
+{"resourceType":"StructureDefinition","id":"us-core-race","text":{"status":"generated","div":"<div xmlns=\"http://www.w3.org/1999/xhtml\"><table border=\"0\" cellpadding=\"0\" cellspacing=\"0\" style=\"border: 0px #F0F0F0 solid; font-size: 11px; font-family: verdana; vertical-align: top;\"><tr style=\"border: 1px #F0F0F0 solid; font-size: 11px; font-family: verdana; vertical-align: top;\"><th sty ...
```
ditto for this and any other repeated SD files.

chandrashekar-s replied:
Done

bashir2 reviewed (approved):
Thanks @chandrashekar-s for the changes. Almost all of my comments are minor/questions. I think the only large ones is asking if we can reduce the number StructureDefinition files from US-Core that are duplicated. Please feel free to merge this once you address/resolve the remaining comments.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMapperFhirContexts.java:
```diff
@@ -0,0 +1,178 @@
+   * Similar to {@link #contextFor(FhirVersionEnum, String)} but the custom profiles are loaded from
+   * the classpath name passed via structureDefinitionsClasspath.
+   */
+  public synchronized FhirContext contextForFromClasspath(
```
Done

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMapperFhirContexts.java:
```diff
@@ -0,0 +1,178 @@
+   * {@code structureDefinitionsPath} path.
+   *
+   * @param context the context into which the structure definitions are loaded
+   * @param structureDefinitionsPath the path containing the custom structure definitions
```
Done

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMappingProvider.java:
```diff
@@ -0,0 +1,274 @@
+   * https://github.com/google/fhir-data-pipes/issues/980
+   *
+   * @param context The context to which the profiles are added.
+   * @param structureDefinitionsPath the path containing the list of structure definitions to be
```
Done

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMappingProvider.java:
```diff
@@ -0,0 +1,274 @@
+      Preconditions.checkNotNull(url, "The url must not be null");
+      if (context.getResourceTypes().contains(type)) {
+        // Add or replace the url for the resource type with a lexicographically smaller one. This
+        // makes sure that for a given set of profiles, it always the chooses the same profile.
```
done

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java:
```diff
@@ -77,20 +77,27 @@ public StructureDefinitions(FhirContext context) {
+      } else {
+        // Remove the leading hash (#) to get the referenced type.
+        String referencedType = parent.getContentReference().substring(1);
+        // Find the actual type to use.
```
Moved the comment to the `getParentDefinition`

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/BasePipelineOptions.java:
```diff
@@ -35,4 +36,32 @@ public interface BasePipelineOptions extends PipelineOptions {
+      "Similar to the structureDefinitionsDir, this path should also contain the structure"
+          + " definition files for any custom profiles that needs to be supported. But the"
+          + " difference is that, this is the classpath name instead of a directory name and should"
+          + " always start with a '/'. Also, only one of the parameter either"
```
done

chandrashekar-s commented on pipelines/controller/config/application.yaml:
```diff
@@ -134,6 +134,24 @@ fhirdata:
+  # Similar to the structureDefinitionsDir, this path should also contain the structure definition
+  # files for any custom profiles that needs to be supported. But the difference is that this is the
+  # classpath name instead of a directory name and should always start with a '/'. Also, only one of
+  # the parameter either structureDefinitionsDir or structureDefinitionsClasspath can be configured,
```
done

chandrashekar-s commented on bunsen/bunsen-avro/src/test/resources/definitions-r4/StructureDefinition-sdc-questionnaireresponse.json:
```diff
@@ -1,44 +1,64 @@
 {
   "resourceType": "StructureDefinition",
-  "id": "us-core-questionnaireresponse",
+  "id": "sdc-questionnaireresponse",
```
These additional artifacts have been removed now.

chandrashekar-s commented on pipelines/common/src/test/java/com/google/fhir/analytics/AvroConversionUtilTest.java:
```diff
@@ -23,52 +23,83 @@
+            .getResource("definitions-r4/StructureDefinition-us-core-patient.json");
+    File file = Paths.get(resourceURL.toURI()).toFile();
+
+    ProfileMapperFhirContexts.getInstance().deRegisterFhirContexts(FhirVersionEnum.R4);
```
Resolving this for now and added the comments in the code.

chandrashekar-s commented on bunsen/bunsen-uscore-resources/src/main/resources/definitions-r4/StructureDefinition-us-core-encounter.json:
```diff
@@ -2472,7 +2472,6 @@
             "code": "Reference",
             "targetProfile": [
               "http://hl7.org/fhir/us/core/StructureDefinition/us-core-condition-problems-health-concerns",
-              "http://hl7.org/fhir/us/core/StructureDefinition/us-core-condition-encounter-diagnosis",
```
This has been addressed
